File: apps/desktop/src/renderer/src/lib/api/types/knowledge.ts
/**
 * Local knowledge base (ChromaDB-backed RAG).
 *
 * Mirrors the FastAPI schemas in `services/agent-engine/app/schemas`.
 */

// ---- Knowledge base (RAG) -------------------------------------------------

/** Response from `GET /knowledge/status`. */
export interface KnowledgeStatus {
  available: boolean;
  engine: string;
  embedding: string;
  path: string | null;
  collection: string;
  document_count: number;
  chunk_count: number;
  message: string | null;
}

/** A knowledge-base document (from `GET /knowledge/documents`). */
export interface KnowledgeDocument {
  id: number;
  title: string;
  file_name: string;
  source_path: string;
  collection_name: string;
  chunk_count: number;
  created_at: string;
  updated_at: string;
}

/** Body for `POST /knowledge/documents/import`. */
export interface KnowledgeImportBody {
  title: string;
  file_name: string;
  content: string;
  file_type?: "txt" | "md";
  source_path?: string;
}

/** A single search hit (from `POST /knowledge/search`). */
export interface KnowledgeSearchResult {
  text: string;
  document_id: number | null;
  chunk_index: number | null;
  title: string | null;
  file_name: string | null;
  distance: number | null;
}

/** Response from `POST /knowledge/search`. */
export interface KnowledgeSearchResponse {
  ok: boolean;
  results: KnowledgeSearchResult[];
  message: string | null;
}


/** Response from `POST /knowledge/sync-agents`. */
export interface KnowledgeSyncResult {
  ok: boolean;
  agents_synced: number;
  files_written: number;
  message: string;
}

File: services/agent-engine/app/schemas/knowledge.py
"""Schemas for the knowledge base (RAG) endpoints."""

from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

KnowledgeFileType = Literal["txt", "md"]


class KnowledgeStatusResponse(BaseModel):
    available: bool
    engine: str = "chromadb (local persistent)"
    embedding: str
    path: Optional[str] = None
    collection: str
    document_count: int = 0
    chunk_count: int = 0
    message: Optional[str] = None


class KnowledgeDocumentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    file_name: str
    source_path: str
    collection_name: str
    chunk_count: int
    created_at: datetime
    updated_at: datetime


class KnowledgeImportRequest(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    file_name: str = Field(min_length=1)
    content: str = Field(min_length=1)
    file_type: KnowledgeFileType = "txt"
    source_path: str = ""


class KnowledgeSearchRequest(BaseModel):
    query: str = Field(min_length=1)
    top_k: Optional[int] = Field(default=None, ge=1, le=20)


class SearchResultItem(BaseModel):
    text: str
    document_id: Optional[int] = None
    chunk_index: Optional[int] = None
    title: Optional[str] = None
    file_name: Optional[str] = None
    distance: Optional[float] = None


class KnowledgeSearchResponse(BaseModel):
    ok: bool
    results: list[SearchResultItem] = Field(default_factory=list)
    message: Optional[str] = None


class DeleteResponse(BaseModel):
    ok: bool


class KnowledgeSyncResponse(BaseModel):
    """Result of mirroring the KB into OpenClaw agent memories."""

    ok: bool
    agents_synced: int = 0
    files_written: int = 0
    message: str = ""
